src/dirac_cwl_proto/metadata/core.py

```diff
@@ -0,0 +1,378 @@
+
+from pydantic import BaseModel, ConfigDict, Field
+
+logger = logging.getLogger(__name__)
+
+# TypeVar for generic class methods
+T = TypeVar("T", bound="TaskDescriptor")
+
+
+class MetadataProcessor(ABC):
+    """Abstract base class for metadata processing hooks.
+
+    This class defines the interface for pre/post processing operations
+    that can be performed during job execution.
+    """
+
+    @abstractmethod
+    def pre_process(self, job_path: Path, command: List[str]) -> List[str]:
+        """Pre-process job inputs and command.
+
+        Parameters
+        ----------
+        job_path : Path
+            Path to the job working directory.
+        command : List[str]
+            The command to be executed.
+
+        Returns
+        -------
+        List[str]
+            Modified command list.
+        """
+        pass
+
+    @abstractmethod
+    def post_process(self, job_path: Path) -> bool:
+        """Post-process job outputs.
+
+        Parameters
+        ----------
+        job_path : Path
+            Path to the job working directory.
+
+        Returns
+        -------
+        bool
+            True if post-processing succeeded, False otherwise.
+        """
+        pass
+
+
+class DataCatalogInterface(ABC):
+    """Abstract interface for data catalog operations."""
+
+    @abstractmethod
+    def get_input_query(self, input_name: str, **kwargs: Any) -> Union[Path, List[Path], None]:
+        """Generate input data query.
+
+        Parameters
+        ----------
+        input_name : str
+            Name of the input parameter.
+        **kwargs : Any
+            Additional query parameters.
+
+        Returns
+        -------
+        Union[Path, List[Path], None]
+            Path(s) to input data or None if not found.
+        """
+        pass
+
+    @abstractmethod
+    def get_output_query(self, output_name: str) -> Optional[Path]:
+        """Generate output data path.
+
+        Parameters
+        ----------
+        output_name : str
+            Name of the output parameter.
+
+        Returns
+        -------
+        Optional[Path]
+            Path where output should be stored or None.
+        """
+        pass
+
+    def store_output(self, output_name: str, src_path: str) -> None:
+        """Store output in the data catalog.
+
+        Parameters
+        ----------
+        output_name : str
+            Name of the output parameter.
+        src_path : str
+            Source path of the output file.
+        """
+        output_path = self.get_output_query(output_name)
+        if not output_path:
+            raise RuntimeError(f"No output path defined for {output_name}")
+
+        output_path.mkdir(exist_ok=True, parents=True)
+        dest = output_path / Path(src_path).name
+        Path(src_path).rename(dest)
+        logger.info(f"Output {output_name} stored in {dest}")
+
+
+class BaseMetadataModel(BaseModel, DataCatalogInterface, MetadataProcessor):
```

Comment:
Here we probably want to change design to the composition instead of inheritance for DataCatalogInterface to allow its reimplementation as `MyDataCatalogInterface`. I won't change it right now to keep its methods in the models, but this would be something to consider.